assets/js/modules/analytics/datastore/accounts.js

```diff
@@ -80,12 +83,97 @@ export const actions = {
 			type: RECEIVE_ACCOUNTS_PROPERTIES_PROFILES_FAILED,
 		};
 	},
+
+	/**
+	 * Creates a new Analytics account.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} args              Argument params.
+	 * @param {string} args.accountName  Google Analytics account name.
+	 * @param {string} args.propertyName Google Analytics property name.
+	 * @param {string} args.profileName  Google Analytics profile name.
+	 * @param {string} args.timezone     Google Analytics timezone.
+	 * @return {Function} Generator function action.
+	 */
+	*createAccount( { accountName, propertyName, profileName, timezone } ) {
+		invariant( accountName, 'accountName is required.' );
+		invariant( propertyName, 'propertyName is required.' );
+		invariant( profileName, 'profileName is required.' );
+		invariant( timezone, 'timezone is required.' );
+
+		try {
+			const accountTicket = yield actions.fetchCreateAccount( { accountName, propertyName, profileName, timezone } );
+			return actions.receiveCreateAccount( { accountTicket } );
+		} catch ( error ) {
+			// TODO: Implement an error handler store or some kind of centralized
+			// place for error dispatch...
+			return actions.receiveCreateAccountFailed( { accountName, error } );
+		}
+	},
+
+	*fetchCreateAccount( { accountName, propertyName, profileName, timezone } ) {
+		yield {
+			type: CREATE_ACCOUNT_STARTED,
+		};
+
+		return {
+			payload: { accountName, propertyName, profileName, timezone },
+			type: FETCH_CREATE_ACCOUNT,
+		};
+	},
+
+	/**
+	 * Adds a account ticket to process.
+	 *
+	 * Adds the newly-created account ticket to the data store.
+	 *
+	 * @since n.e.x.t
+	 * @private
+	 *
+	 * @param {Object} args               Argument params.
+	 * @param {Object} args.accountTicket Google Analytics create account ticket object.
+	 */
+	receiveCreateAccount( { accountTicket } ) {
+		invariant( accountTicket, 'accountTicket is required.' );
+
+		// Once we have an account ticket, redirect the user to accept the Terms of Service.
+		const { id } = accountTicket;
+		if ( id ) {
+			document.location = `https://analytics.google.com/analytics/web/?provisioningSignup=false#management/TermsOfService/?api.accountTicketId=${ id }`;
+		}
```

Comment:
Strictly we'd also need an action creator here, and we'd need to set the request flag to false. However, since we do a redirect here which will destroy the context, I don't think it's a big deal here.

@tofumatt We talked about this before, and I think it makes sense to have that redirect in the datastore, since it must always happen. Let me know if you have a preference to potentially include this in the datastore in another way.

Potentially it would be cleaner to have a `redirectToTOS` action so that at least the create-account part can be tested in isolation from that.

Reply:
@tofumatt @aaemnnosttv @adamsilverstein I'm not aware of any downside of using `window.location.assign()` instead of `document.location` - just pinging you all in case you know of any reason why we shouldn't be using it.

Reply:
I can add an inline comment to note that this is so we can test the location has been set. I think I can just use `location.assign` here (dropping the `window` part and remove the linting comments.

Reply:
Not sure a comment is necessary. If there's no trade-off here, maybe we should be using this throughout our codebase for the future, since generally a function like this is better for testing.

Reply:
@adamsilverstein you're getting the eslint error because we don't allow use of `window` - use `global` instead.

I agree that we should use methods on `location` for changes.  I did some looking to see if there's any reason why we shouldn't use it and couldn't find anything. Assigning a URL to `location` is also valid per [the HTML standard](https://html.spec.whatwg.org/multipage/history.html#the-location-interface) but probably harder to test.

Reply:
I don't like the inclusion of the `window.location` call here because:

1. It makes the data store more difficult to test; we need to mock `window.location` to test a data store, which doesn't make a whole lot of sense to me.
2. Separation of concerns: adjusting the location of the page isn't a datastore concern; it's a component/UI concern. This is logic that is better-placed in a `useEffect()` call of the component that dispatches whatever action triggers this behaviour. If the `id` is received then we perform this redirect, but in the UI (eg the component), not in the datastore.

Our datastore should not ever be concerned with routing or page URL in such a _direct_ way. I think it would be better to exclude this from the PR and incorporate this logic into the component that dispatches these actions.

-----

If we want a more generic way to handle redirects I'm even fine with a `setRedirect( url )` action and a `getRedirect()` selector in a high-level component that handles this at an app-wide level (some component high up in the component tree whose only job it is is to select that data and redirect with `window.location` when it happens). But I don't think this should be done here.

Reply:
Thanks for the feedback @tofumatt, that's what I had pinged you about before.

That makes sense to me, it's a bit odd that the datastore handles this, but since it destroys the page context anyway I wasn't sure what's recommended here.

Based on this, I suggest we modify `receiveCreateAccount` to return an action creator object `{ payload: { accountTicket }, type: RECEIVE_CREATE_ACCOUNT }`, and then add a `RECEIVE_CREATE_ACCOUNT` case in the reducer that sets a state property `accountTicketTOSURL` or `accountTicketTermsOfServiceURL` (the latter sounds better, but also is _very_ long, so I'm not sure). A selector `getAccountTicketTOSURL()` or `getAccountTicketTermsOfServiceURL()` would then allow to get that URL once set, so that the component can take care of redirecting there. (Note that the selector/state property should be more specific to what this URL is - it shouldn't involve "redirect" in its name, since that has been a kind of "reserved" term for OAuth redirect URL in our plugin, so describing what this URL is for I think is more accurate.)

Reply:
Ok, thanks for the feedback @tofumatt I agree with your points and was originally planning to have the redirect in the component itself based on the returned value. I will make it so.

Reply:
@felixarntz I changed this as you suggested in https://github.com/google/site-kit-wp/pull/1390/commits/d4f1bc132be3e220d3b73747b1d7ed8b7f6b9c99 - one other option would be to set only the ticket or id, then construct the url in the client. Sticking with setting the URL here and I'll add the redirect logic to the component.

Reply:
I think dealing with the URL in the datastore keeps that functionality encapsulated there. From component POV, the only thing it needs is the URL, not the account ticket ID, that's why I thought exposing the full URL would make more sense there.

Reply:
Thanks for the fixes on this 👍